### PR TITLE
ApiManagement - Refresh cmdlets based on autorest .net client 2018-01-01

### PR DESCRIPTION
**Pull request by solankisamir**
## Description

- Migrated the apimanagement powershell cmdlets from hyak based .net nuget to autorest based nuget
- Added support for ApiVersions, ApiReleases and ApiRevisions
- Added suppport for ServiceFabric Backend
- Added support for Application Insights Logger
- Added support for `Basic` sku
- Added support for installing Certificates issued by private CA as Root or CA
- Added support for accepting Custom SSL certificates via KeyVault
- Added support for MSI
- Added support for accepting Policies via Url

Cmdlet design review https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/35

This PR fixes the following Issues
https://github.com/Azure/azure-powershell/issues/5812
https://github.com/Azure/azure-powershell/issues/5021
https://github.com/Azure/azure-powershell/issues/5724

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by MiYanni** [comment]
@azuresdkci Retest this please

**Comment by solankisamir** [comment]
@cormacpayne @markcowl can you take a look at this PR? We have customers waiting on this functionality and bug fixes. Thanks

**Comment by solankisamir** [comment]
@markcowl or @cormacpayne can you can take a look at this PR?

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/RemoveAzureApiManagementApiVersionSet.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        [ValidateNotNullOrEmpty]
+        public PsApiManagementContext Context { get; set; }
+
+        [Parameter(
+            ParameterSetName = ByValueParameterSet,
+            ValueFromPipeline = true,
+            Mandatory = true,
+            HelpMessage = "Instance of PsApiManagementApiVersionSet. This parameter is required.")]
+        [ValidateNotNullOrEmpty]
+        public PsApiManagementApiVersionSet ApiVersionSetObject { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/RemoveAzureApiManagementApiVersionSet.cs]: @solankisamir this should be renamed to `InputObject`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApi.cs [hunk 2]
```diff
@@ -77,29 +90,41 @@ public class SetAzureApiManagementApi : AzureApiManagementCmdletBase
...
             HelpMessage = "Subscription key query string parameter name. This parameter is optional. Default value is $null.")]
         public String SubscriptionKeyQueryParamName { get; set; }
+        
+        [Parameter(
+            ParameterSetName = ByValueParameterSet,
+            ValueFromPipeline = true,
+            Mandatory = true,
+            HelpMessage = "Instance of PsApiManagementApi. This parameter is required.")]
+        [ValidateNotNullOrEmpty]
+        public PsApiManagementApi ApiObject { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApi.cs]: @solankisamir this should be renamed to `InputObject`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApiVersionSet.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        [ValidateNotNullOrEmpty]
+        public PsApiManagementContext Context { get; set; }
+
+        [Parameter(
+            ParameterSetName = ByValueParameterSet,
+            ValueFromPipeline = true,
+            Mandatory = true,
+            HelpMessage = "Instance of PsApiManagementApiVersionSet. This parameter is required.")]
+        [ValidateNotNullOrEmpty]
+        public PsApiManagementApiVersionSet ApiVersionSetObject { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApiVersionSet.cs]: @solankisamir this should be renamed to `InputObject`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementLogger.cs [hunk 4]
```diff
@@ -14,41 +14,48 @@
...
             Mandatory = false,
             HelpMessage = "EventHub Connection String with Send Policy Rights. This parameter is optional.")]
         public String ConnectionString { get; set; }
+       
+        [Parameter(            
+            ValueFromPipelineByPropertyName = true,
+            Mandatory = false,
+            HelpMessage = "Instrumentation Key of the application Insights. This parameter is optional.")]
+        [ValidateNotNullOrEmpty]
+        public String InstrumentationKey { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementLogger.cs]: @solankisamir do we need to put this parameter in a separate parameter set from `ConnectionString` like we did for the `New-*` cmdlet? This will avoid allowing users to provide both in the same set

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/UpdateAzureApiManagementApiRelease.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            Mandatory = false,
+            HelpMessage = "Api Release Notes. This parameter is optional.")]
+        public String Note { get; set; }
+
+        [Parameter(
+            ParameterSetName = ByValueParameterSet,
+            ValueFromPipelineByPropertyName = true,
+            Mandatory = false,
+            HelpMessage = "Instance of type Microsoft.Azure.Commands.ApiManagement.ServiceManagement.Models.PsApiManagementApiRelease.")]
+        public PsApiManagementApiRelease ApiRelease { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/UpdateAzureApiManagementApiRelease.cs]: @solankisamir this should be renamed to `InputObject`, should be mandatory in its parameter set, and should have `ValueFromPipeline = true`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/packages.config [hunk 6]
```diff
@@ -1,5 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="AutoMapper" version="6.0.2" targetFramework="net452" />
-  <package id="Microsoft.Azure.Management.ApiManagement" version="3.4.0-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.KeyVault.Core" version="1.0.0" targetFramework="net45" />
+  <package id="Microsoft.Azure.Management.ApiManagement" version="4.0.2-preview" targetFramework="net452" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.11" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/packages.config]: @solankisamir same comment as in the `.csproj` file -- you can remove the files added (except `WindowsAzure.Storage`) as they are referenced elsewhere

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.Test/Commands.ApiManagement.Test.csproj [hunk 7]
```diff
@@ -84,8 +83,20 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\..\packages\Microsoft.IdentityModel.Clients.ActiveDirectory.2.28.3\lib\net45\Microsoft.IdentityModel.Clients.ActiveDirectory.WindowsForms.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Rest.ClientRuntime.2.3.11\lib\net452\Microsoft.Rest.ClientRuntime.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime.Azure, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.Test/Commands.ApiManagement.Test.csproj]: @solankisamir you can remove the reference to `Microsoft.Rest.ClientRuntime` and `Microsoft.Rest.ClientRuntime.Azure` from this file as they are referenced elsewhere

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj [hunk 8]
```diff
@@ -64,10 +67,23 @@
     <Reference Include="Microsoft.Data.Services.Client, Version=5.8.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
       <HintPath>..\..\..\packages\Microsoft.Data.Services.Client.5.8.2\lib\net40\Microsoft.Data.Services.Client.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj]: @solankisamir same comment about removing the added packages

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands/SetAzureApiManagement.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    [Cmdlet(VerbsCommon.Set, "AzureRmApiManagement", SupportsShouldProcess = true), OutputType(typeof(PsApiManagement))]
+    public class SetAzureApiManagement : AzureApiManagementCmdletBase
+    {
+        [Parameter(
+            ValueFromPipeline = true,
+            Mandatory = true,
+            HelpMessage = "The ApiManagement instance.")]
+        [ValidateNotNull]
+        public PsApiManagement ApiManagement { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands/SetAzureApiManagement.cs]: @solankisamir this should be renamed to `InputObject`

Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands/SetAzureApiManagement.cs]: @solankisamir also, we should look at adding another parameter set containing the identity properties of an `ApiManagement` object that the user could provide to this cmdlet so they don't always have to provide the `ApiManagement` object

File: src/ResourceManager/ApiManagement/Commands.SMAPI.Test/ScenarioTests/ApiManagementTests.ps1 [hunk 10]
```diff
@@ -175,10 +175,12 @@ function Api-ImportExportWadlTest {
...
         Assert-AreEqual $path $api.Path
 
+        # commented as powershell test framework on running test in playback mode, throws 403, as the exported link of file
+        # gets expired
         # export api to pipline
-        $result = Export-AzureRmApiManagementApi -Context $context -ApiId $wadlApiId -SpecificationFormat Wadl
+        # $result = Export-AzureRmApiManagementApi -Context $context -ApiId $wadlApiId -SpecificationFormat Wadl
 
-        Assert-True {$result -like '*<doc title="Yahoo News Search">Yahoo News Search API</doc>*'}
+        # Assert-True {$result -like '*<doc title="Yahoo News Search">Yahoo News Search API</doc>*'}
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.SMAPI.Test/ScenarioTests/ApiManagementTests.ps1]: @solankisamir it looks like there's a lot of commented-out asserts in this file -- any reason why?

File: src/ResourceManager/ApiManagement/Commands.SMAPI.Test/packages.config [hunk 11]
```diff
@@ -4,22 +4,25 @@
...
-  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.6.7-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.8.1" targetFramework="net452" />
   <package id="Microsoft.Bcl" version="1.1.9" targetFramework="net45" />
   <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net45" />
   <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net45" />
   <package id="Microsoft.IdentityModel.Clients.ActiveDirectory" version="2.28.3" targetFramework="net45" />
   <package id="Microsoft.Net.Http" version="2.2.28" targetFramework="net45" />
-  <package id="Microsoft.Rest.ClientRuntime.Azure.Authentication" version="2.2.9-preview" targetFramework="net45" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.11" targetFramework="net452" />
+  <package id="Microsoft.Rest.ClientRuntime.Azure" version="3.3.12" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.SMAPI.Test/packages.config]: @solankisamir same comment about removing the additional packages

File: src/StackAdmin/AzureRM/AzureRM.psd1 [hunk 12]
```diff
@@ -1,100 +1,103 @@
-﻿#
```
Comment (cormacpayne) [inline src/StackAdmin/AzureRM/AzureRM.psd1]: @solankisamir please revert the changes made in this file, both of the `AzureRM.psm1` files, and `AzureStack.psm1`

File: tools/StaticAnalysis/Exceptions/BreakingChangeIssues.csv [hunk 13]
```diff
@@ -1 +1,2 @@
-"AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
\ No newline at end of file
+"AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
+"Microsoft.Azure.Commands.ApiManagement.ServiceManagement.dll","Microsoft.Azure.Commands.ApiManagement.ServiceManagement.Commands.SetAzureApiManagementPolicy","Set-AzureRmApiManagementPolicy","0","2050","A validate set has been added for parameter 'Format' for cmdlet 'Set-AzureRmApiManagementPolicy'.","Remove the validate set from parameter 'Format'."
```
Comment (cormacpayne) [inline tools/StaticAnalysis/Exceptions/BreakingChangeIssues.csv]: @solankisamir can you confirm that this isn't a breaking change? If a user were to provide values outside of the values in the `ValidateSet`, would the cmdlet throw an exception?

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj [hunk 14]
```diff
@@ -51,9 +51,40 @@
       <HintPath>..\..\..\packages\AutoMapper.6.0.2\lib\net45\AutoMapper.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.ApiManagement, Version=3.4.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <SpecificVersion>False</SpecificVersion>
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.ApiManagement.3.4.0-preview\lib\net45\Microsoft.Azure.Management.ApiManagement.dll</HintPath>
+    <Reference Include="Microsoft.Azure.KeyVault.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj]: fixed

---
In reply to: [187676809](https://github.com/Azure/azure-powershell/pull/6189#discussion_r187676809) [](ancestors = 187676809)

File: src/ResourceManager/ApiManagement/Commands.SMAPI.Test/ScenarioTests/ApiManagementTests.ps1 (hunk 10 above)
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.SMAPI.Test/ScenarioTests/ApiManagementTests.ps1]: The output of the commandlet is an Uri string with time bound token. When we rerun the test in Playback mode, the commandlet throws exception, as the Uri is no longer valid.
All the commented Commandlet are the one testing `Export-AzureRmApiManagementApi`


---
In reply to: [188128363](https://github.com/Azure/azure-powershell/pull/6189#discussion_r188128363) [](ancestors = 188128363)

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands/SetAzureApiManagement.cs (hunk 9 above)
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands/SetAzureApiManagement.cs]: I will take it as feedback for the next update to the commandlet.

---
In reply to: [188127907](https://github.com/Azure/azure-powershell/pull/6189#discussion_r188127907) [](ancestors = 188127907)

**Comment by solankisamir** [comment]
@cormacpayne I have incorporated the review comments. Can you take a look again?

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj [hunk 15]
```diff
@@ -24,7 +24,7 @@
     <WarningLevel>4</WarningLevel>
     <RunCodeAnalysis>true</RunCodeAnalysis>
     <UseVSHostingProcess>true</UseVSHostingProcess>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj]: @solankisamir why were these changed to `false`?

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj (hunk 14 above)
[the remark this answers is not in the document]
Comment (cormacpayne) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj]: @solankisamir please remove `Microsoft.Azure.KeyVault.Core` and `Newtonsoft.Json` references from this file

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj (hunk 15 above)
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj]: Marking a parameter/cmdlet as Obsolete was causing Build Errors, as they come as Warning.

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/NewAzureApiManagementLogger.cs [hunk 16]
```diff
@@ -38,26 +41,37 @@ public class NewAzureApiManagementLogger : AzureApiManagementCmdletBase
...
             HelpMessage = "EventHub Entity name. This parameter is required.")]
         [ValidateNotNullOrEmpty]
         public String Name { get; set; }
 
         [Parameter(
+            ParameterSetName = EventHubLoggerSet,
             ValueFromPipelineByPropertyName = true,
             Mandatory = true,
-            HelpMessage = "EventHub Connection String with Send Policy Rights. This parameter is required.")]
+            HelpMessage = "EventHub Connection String with Send Policy Rights. This parameter is optional.")]
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/NewAzureApiManagementLogger.cs]: @solankisamir "This parameter is optional" even though `Mandatory = true`

Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/NewAzureApiManagementLogger.cs]: will fix it.

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/RemoveAzureApiManagementApiRelease.cs [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        #region Parameter Set Names
+
+        private const string ByApiReleaseIdParameterSet = "ByApiReleaseId";
+        private const string ByInputObjectParameterSet = "ByInputObject";
+
+        #endregion
+
+        [Parameter(
+            ParameterSetName = ByApiReleaseIdParameterSet,
+            ValueFromPipeline = true,
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/RemoveAzureApiManagementApiRelease.cs]: @solankisamir `ValueFromPipelineByPropertyName = true`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj (hunk 15 above)
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands.ApiManagement.ServiceManagement.csproj]: Marking a parameter/cmdlet as Obsolete was causing Build Errors, as they come as Warning.

---
In reply to: [188763417](https://github.com/Azure/azure-powershell/pull/6189#discussion_r188763417) [](ancestors = 188763417)

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApiVersionSet.cs [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        [Parameter(
+            ParameterSetName = ExpandedParameterSet,
+            ValueFromPipelineByPropertyName = true,
+            Mandatory = false,
+            HelpMessage = "The Query value which will contain the versioning information. " +
+            "If versioning Scheme Query is choosen, then this value must be specified.")]
+        public String QueryName { get; set; }
+
+        [Parameter(
+            ParameterSetName = ExpandedParameterSet,
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApiVersionSet.cs]: @solankisamir all of the parameters that the user can update on a version set should be in _both_ parameter sets, or else you are setting `null` for all properties when the user provides `InputObject`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApiRevision.cs [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+            string id = apiId.ApiRevisionIdentifier(apiRevision);
+
+            if (ShouldProcess(id, Resources.SetApiRevision))
+            {
+                Client.ApiSet(
+                    resourcegroupName,
+                    serviceName,
+                    id,
+                    Name,
+                    Description,
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/SetAzureApiManagementApiRevision.cs]: @solankisamir make sure that the parameters that the user can set for a revision are in _both_ parameter sets so the user can update properties when providing an `InputObject`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/UpdateAzureApiManagementApiRelease.cs [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+            ValueFromPipelineByPropertyName = true,
+            ParameterSetName = ExpandedParameterSet,
+            Mandatory = true,
+            HelpMessage = "Identifier of existing API. This parameter is required.")]
+        [ValidateNotNullOrEmpty]
+        public String ApiId { get; set; }
+                
+        [Parameter(
+            ValueFromPipelineByPropertyName = true,
+            ParameterSetName = ExpandedParameterSet,
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/UpdateAzureApiManagementApiRelease.cs]: @solankisamir make sure that these parameters that can be updated for a release are in _both_ parameter sets so a user can update the properties when providing an `InputObject`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/packages.config [hunk 21]
```diff
@@ -1,5 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="AutoMapper" version="6.0.2" targetFramework="net452" />
-  <package id="Microsoft.Azure.Management.ApiManagement" version="3.4.0-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.KeyVault.Core" version="1.0.0" targetFramework="net45" />
+  <package id="Microsoft.Azure.Management.ApiManagement" version="4.0.2-preview" targetFramework="net452" />
+  <package id="Newtonsoft.Json" version="6.0.8" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/packages.config]: @solankisamir please remove `Microsoft.Azure.KeyVault.Core` and `Newtonsoft.Json` from this file

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj [hunk 22]
```diff
@@ -22,7 +22,7 @@
     <DefineConstants>TRACE;DEBUG;CODE_ANALYSIS</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj]: @solankisamir same comment about why this was changed to `false`

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj [hunk 23]
```diff
@@ -50,9 +50,12 @@
       <HintPath>..\..\..\packages\AutoMapper.6.0.2\lib\net45\AutoMapper.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.ApiManagement, Version=3.4.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <SpecificVersion>False</SpecificVersion>
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.ApiManagement.3.4.0-preview\lib\net45\Microsoft.Azure.Management.ApiManagement.dll</HintPath>
+    <Reference Include="Microsoft.Azure.KeyVault.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj]: @solankisamir same comment about removing `Microsoft.Azure.KeyVault.Core` and `Newtonsoft.Json` from this file

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj (hunk 22 above)
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/Commands.ApiManagement.csproj]: Marking a parameter/cmdlet obsolete was throwing warning, which were being treated as Build errors.

---
In reply to: [188767844](https://github.com/Azure/azure-powershell/pull/6189#discussion_r188767844) [](ancestors = 188767844)

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementApiRelease.md [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+## DESCRIPTION
+
+The **New-AzureRmApiManagementApiRelease** cmdlet creates an API Release for an API Revision in API Management context.
+
+## EXAMPLES
+
+### Example 1: Create an API Release for an API Revision
+```powershell
+PS C:\>$ApiMgmtContext = New-AzureRmApiManagementContext -ResourceGroupName "Api-Default-WestUS" -ServiceName "contoso"
+PS C:\>New-AzureRmApiManagementApiRelease -Context $ApiMgmtContext -ApiId "echo-api" -ApiRevision "2" -ReleaseId "echo-api-release" -Note "Releasing version 2 of the echo-api to public"
```
Comment (cormacpayne) [inline src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementApiRelease.md]: @solankisamir please make sure your examples have a sample output from when the command is executed to give users a better idea of what they can expect to be returned from the cmdlet invocation

Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementApiRelease.md]: can you provide a sample of what you are saying? I am not clear.

---
In reply to: [188768556](https://github.com/Azure/azure-powershell/pull/6189#discussion_r188768556) [](ancestors = 188768556)

Comment (cormacpayne) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementApiRelease.md]: @solankisamir see the examples for [`Get-AzureRmResource`](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmResource.md#examples) -- all of the examples of a sample output of what the user would see when executing the cmdlet

**Comment by cormacpayne** [comment]
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/853/
